src/lava/magma/runtime/message_infrastructure/message_infrastructure/csrc/core/message_infrastructure_logging.h

```diff
@@ -5,40 +5,164 @@
+
+#if defined(MSG_LOG_FILE_ENABLE)
+#define DEBUG_LOG_PRINT(_level, _fmt, ...) do { \
+    int length = 0; \
+    char *log_data = reinterpret_cast<char *> \
+                      (malloc(sizeof(char)*MAX_SIZE_PER_LOG_MSG)); \
+    if (log_data != NULL) { \
+      length = std::snprintf(log_data, \
+                             MAX_SIZE_PER_LOG_MSG, _fmt, ## __VA_ARGS__); \
+    } \
+    if (log_data == NULL || length <0) { \
+      GetLogInstance()->LogWrite(LogMsg(std::string(LOG_MSG_SUBSTITUTION), \
+                                        __FILE__, \
+                                        __LINE__, \
+                                        _level)); \
+    } else { \
+      GetLogInstance()->LogWrite(LogMsg(std::string(log_data), \
+                                        __FILE__, \
+                                        __LINE__, \
+                                        _level)); \
+    } \
+    free(log_data); \
+    std::printf("%s", _level); \
+    std::printf(_fmt, ## __VA_ARGS__); \
+} while (0)
+#else
+#define DEBUG_LOG_PRINT(_level, _fmt, ...) do { \
+  std::printf("%s", _level); \
+  std::printf(_fmt, ## __VA_ARGS__); \
+} while (0)
+#endif
+
+#define LAVA_LOG(_module, _fmt, ...) do { \
+  if ((_module) && (MSG_LOG_LEVEL <= LOG_MASK_INFO)) { \
+    DEBUG_LOG_PRINT("[CPP INFO]", _fmt, ## __VA_ARGS__); \
   } \
-}
+} while (0)
 
-#define LAVA_DUMP(_cond, _fmt, ...) { \
-  if ((_cond && DEBUG_MODE)) { \
-    printf(_fmt, ## __VA_ARGS__); \
+#define LAVA_DUMP(_module, _fmt, ...) do { \
+  if (_module && (MSG_LOG_LEVEL <= LOG_MASK_DUMP)) { \
+    DEBUG_LOG_PRINT("[CPP DUMP]", _fmt, ## __VA_ARGS__); \
   } \
-}
+} while (0)
 
-#define LAVA_DEBUG(_cond, _fmt, ...) { \
-  if ((_cond && DEBUG_MODE)) { \
-    printf("[CPP DEBUG]" _fmt, ## __VA_ARGS__); \
+#define LAVA_DEBUG(_module, _fmt, ...) do { \
+  if (_module && (MSG_LOG_LEVEL <= LOG_MASK_DBUG)) { \
+    DEBUG_LOG_PRINT("[CPP DBUG]", _fmt, ## __VA_ARGS__); \
   } \
-}
+} while (0)
 
-#define LAVA_LOG_WARN(_cond, _fmt, ...) { \
-  if ((_cond)) { \
-    printf("[CPP WARNING] %s " _fmt, __FUNCTION__, ## __VA_ARGS__); \
+#define LAVA_LOG_WARN(_module, _fmt, ...) do { \
+  if (_module && (MSG_LOG_LEVEL <= LOG_MASK_WARN)) { \
+    DEBUG_LOG_PRINT("[CPP WARN]", _fmt, ## __VA_ARGS__); \
   } \
-}
+} while (0)
+
+#define LAVA_LOG_ERR(_fmt, ...) do { \
+  DEBUG_LOG_PRINT("[CPP ERRO]", _fmt, ## __VA_ARGS__); \
+} while (0)
+
+namespace message_infrastructure {
+
+enum LogLevel {
+  LOG_MASK_DBUG,
+  LOG_MASK_INFO,
+  LOG_MASK_DUMP,
+  LOG_MASK_WARN,
+  LOG_MASK_ERRO
+};
+
+class LogMsg{
+ public:
+  LogMsg(const std::string &msg_data,
+         const char *log_file,
+         const int log_line,
+         const char *log_level);
+  std::string GetEntireLogMsg(int pid);
+
+ private:
+  std::string msg_time_;
+  std::string msg_data_;
+  std::string msg_level_;
+  std::string msg_file_;
+  int msg_line_ = 0;
+};
+
+class MessageInfrastructureLog {
+ public:
+  MessageInfrastructureLog();
```

Comment:
If you want to implement MessageInfrastructureLog as a singleton pattern, move the constructor func to private. If not, just ignore this comment.

Reply:
it's a singleton pattern, but the instance isn't a static, but a global variable, and it will be constructed while the first call. Not all the time the instance will be constructed, it depends on the MSG_LOG_FILE_ENABLE macro. The private constructor will cause an error while the first call.